designs/2019-expose-rules-to-formatters/readme.MD

```diff
@@ -0,0 +1,137 @@
+- Start Date: 2019-01-16
+- RFC PR: (leave this empty, to be filled in later)
+- Authors: Chris Meyer (@EasyRhinoMSFT)
+
+# Providing Rule Metadata to Formatters
+
+## Summary
+
+This proposal describes a design enhancement that provides formatters with details about the rules that have been executed by ESLint.
+
+## Motivation
+
+Currently, formatters only see the ID of each rule for which a violation was identified, plus an instance-specific description, as properties on each result object. Formatters are not able to access usefule rule metadata, such as category, description, and help URL. Formatters are also not aware of the full set of rules that were run, information that may be useful in some cases.
+
+## Detailed Design
+
+Design Summary
+1. In `cli.js::printResults`, obtain the rules map from the `Engine` object.
+2. Pass the rules map as property in a new, second parameter object to the formatter's exported interface function.
+
+We should use a container object as the parameter, with the rules map as a property, in order to accommodate potential future expansions of the data we pass to formatters. This suggestion was previously made in the discussion of issue [#9841](https://github.com/eslint/eslint/issues/9841).
+
+### Command Line Interface (cli.js) Changes
+The implementation of this feature is very simple and straightfoward. The code location that invokes the formatter's exported interface function already has access to the API it should use to obtain the list of rules, and the output of that function is a Map (dictionary) which is the type we want to pass to the formatter. This dataset includes all executed rules. The call to `Engine.getRules` must be made in the try block because `enginge` may be null during unit testing.
```

Comment:
```suggestion
The implementation of this feature is very simple and straightfoward. The code location that invokes the formatter's exported interface function already has access to the API it should use to obtain the list of rules, and the output of that function is a Map (dictionary) which is the type we want to pass to the formatter. This dataset includes all executed rules. The call to `Engine.getRules` must be made in the try block because `engine` may be null during unit testing.
```